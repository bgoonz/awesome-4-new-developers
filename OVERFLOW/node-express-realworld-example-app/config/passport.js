var passport = require('passport');
var LocalStrategy = require('passport-local').Strategy;
var mongoose = require('mongoose');
var User = mongoose.model('User');

passport.use(new LocalStrategy({
  usernameField: 'user[email]',
  passwordField: 'user[password]'
}, function(email, password, done) {
  User.findOne({email: email}).then(function(user){
    if(!user || !user.validPassword(password)){
      return done(null, false, {errors: {'email or password': 'is invalid'}});
    }

    return done(null, user);
  }).catch(done);
}));

