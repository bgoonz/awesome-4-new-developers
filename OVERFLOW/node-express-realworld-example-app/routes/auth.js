var jwt = require('express-jwt');
var secret = require('../config').secret;

function getTokenFromHeader(req){
  if (req.headers.authorization && req.headers.authorization.split(' ')[0] === 'Token' ||
      req.headers.authorization && req.headers.authorization.split(' ')[0] === 'Bearer') {
    return req.headers.authorization.split(' ')[1];
  }

  return null;
}

var auth = {
  required: jwt({
    secret: secret,
    userProperty: 'payload',
    getToken: getTokenFromHeader
  }),
  optional: jwt({
    secret: secret,
    userProperty: 'payload',
    credentialsRequired: false,
    getToken: getTokenFromHeader
  })
};

module.exports = auth;
