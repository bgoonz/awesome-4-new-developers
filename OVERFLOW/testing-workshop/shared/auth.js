const crypto = require('crypto')
const jwt = require('jsonwebtoken')

const secret = 'secret'

const iterations = process.env.NODE_ENV === 'production' ? 100000 : 1

// seconds/minute * minutes/hour * hours/day * 60 days
const sixtyDaysInSeconds = 60 * 60 * 24 * 60
// to keep our tests reliable, we'll use the requireTime if we're not in production
// and we'll use Date.now() if we are.
const requireTime = Date.now()
const now = () =>
  process.env.NODE_ENV === 'production' ? Date.now() : requireTime

function getSaltAndHash(password) {
  const salt = crypto.randomBytes(16).toString('hex')
  const hash = crypto
    .pbkdf2Sync(password, salt, iterations, 512, 'sha512')
    .toString('hex')
  return {salt, hash}
}

function isPasswordValid(password, {salt, hash}) {
  return (
    hash ===
    crypto.pbkdf2Sync(password, salt, iterations, 512, 'sha512').toString('hex')
  )
}

function getUserToken({id, username}) {
  const issuedAt = Math.floor(now() / 1000)
  return jwt.sign(
    {
      id,
      username,
      iat: issuedAt,
      exp: issuedAt + sixtyDaysInSeconds,
    },
    secret,
  )
}

module.exports = {getSaltAndHash, isPasswordValid, getUserToken, secret}
