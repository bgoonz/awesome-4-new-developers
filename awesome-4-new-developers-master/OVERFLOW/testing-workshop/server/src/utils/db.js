import * as generate from 'til-shared/generate'

const db = {
  users: [],
  posts: [],

  insertUser,
  getUser,
  getUsers,
  updateUser,
  deleteUser,

  insertPost,
  getPost,
  getPosts,
  updatePost,
  deletePost,
}

async function insertUser(user) {
  const newUser = {
    ...user,
    id: generate.id(),
  }
  db.users.push(newUser)
  return newUser
}

async function getUsers(filter) {
  return filter ? db.users.filter(filter) : [...db.users]
}

async function getUser(id) {
  return (await getUsers(u => u.id === id))[0]
}

async function updateUser(id, newInfo) {
  const user = await getUser(id)
  // doing this to make a new copy of the user to avoid subtle bugs
  // that rely on mutation.
  const newUserWithUpdates = Object.assign({}, user, newInfo)
  db.users[db.users.indexOf(user)] = newUserWithUpdates
  return newUserWithUpdates
}

async function deleteUser(id) {
  const user = await getUser(id)
  db.users = db.users.filter(u => u.id !== id)
  return user
}

async function insertPost(post) {
  const newPost = {
    ...post,
    id: generate.id(),
  }
  db.posts.push(newPost)
  return newPost
}

async function getPosts(filter) {
  return filter ? db.posts.filter(filter) : [...db.posts]
}

async function getPost(id) {
  return (await getPosts(t => t.id === id))[0]
}

async function updatePost(id, newInfo) {
  const post = await getPost(id)
  // doing this to make a new copy of the user to avoid subtle bugs
  // that rely on mutation.
  const newPostWithUpdates = Object.assign({}, post, newInfo)
  db.posts[db.posts.indexOf(post)] = newPostWithUpdates
  return newPostWithUpdates
}

async function deletePost(id) {
  const post = await getPost(id)
  db.posts = db.posts.filter(t => t.id !== id)
  return post
}

export default db
