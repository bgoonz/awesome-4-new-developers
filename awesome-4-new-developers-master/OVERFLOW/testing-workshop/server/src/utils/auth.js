import expressJWT from 'express-jwt'
import LocalStrategy from 'passport-local'
import {omit} from 'lodash'
import {
  getSaltAndHash,
  isPasswordValid,
  secret,
  getUserToken,
} from 'til-shared/auth'
import db from './db'

const authMiddleware = {
  required: expressJWT({
    secret,
  }),
  optional: expressJWT({
    secret,
    credentialsRequired: false,
  }),
}

function getLocalStrategy() {
  return new LocalStrategy(async (username, password, done) => {
    let user
    try {
      user = (await db.getUsers(u => u.username === username))[0]
    } catch (error) {
      return done(error)
    }
    if (!user || !isPasswordValid(password, user)) {
      return done(null, false, {
        errors: {'username or password': 'is invalid'},
      })
    }
    return done(null, userToJSON(user))
  })
}

function userToJSON(user) {
  return omit(user, ['exp', 'iat', 'hash', 'salt'])
}

function isPasswordAllowed(password) {
  return password.length > 6 && /\d/.test(password) && /\D/.test(password)
}

export {
  authMiddleware,
  getSaltAndHash,
  userToJSON,
  getLocalStrategy,
  getUserToken,
  isPasswordAllowed,
}
