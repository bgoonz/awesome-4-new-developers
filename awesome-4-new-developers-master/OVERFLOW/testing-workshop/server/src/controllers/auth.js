import passport from 'passport'
import {getSaltAndHash, userToJSON, getUserToken} from '../utils/auth'
import db from '../utils/db'

const authUserToJSON = user => ({
  ...userToJSON(user),
  token: getUserToken(user),
})

async function register(req, res) {
  const {username, password} = req.body
  if (!username) {
    return res.status(422).json({errors: {username: `can't be blank`}})
  }

  if (!password) {
    return res.status(422).json({errors: {password: `can't be blank`}})
  }
  const existingUser = (await db.getUsers(u => u.username === username))[0]
  if (existingUser) {
    return res.status(422).json({errors: {username: 'taken'}})
  }
  const newUser = await db.insertUser({
    username,
    ...getSaltAndHash(password),
  })
  return res.json({user: authUserToJSON(newUser)})
}

async function login(req, res, next) {
  if (!req.body.username) {
    return res.status(422).json({errors: {username: `can't be blank`}})
  }

  if (!req.body.password) {
    return res.status(422).json({errors: {password: `can't be blank`}})
  }
  const {user, info} = await authenticate(req, res, next)

  if (user) {
    return res.json({user: authUserToJSON(user)})
  } else {
    return res.status(422).json(info)
  }
}

function authenticate(req, res, next) {
  return new Promise((resolve, reject) => {
    passport.authenticate('local', {session: false}, (err, user, info) => {
      if (err) {
        reject(err)
      } else {
        resolve({user, info})
      }
    })(req, res, next)
  })
}

async function me(req, res) {
  if (req.user) {
    return res.json({user: authUserToJSON(req.user)})
  } else {
    return res.status(404).send()
  }
}

export {me, login, register}
