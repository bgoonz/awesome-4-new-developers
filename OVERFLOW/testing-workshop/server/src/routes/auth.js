import {authMiddleware} from '../utils/auth'
import * as authController from '../controllers/auth'

function setupAuthRoutes(router) {
  router.post('/register', authController.register)
  router.post('/login', authController.login)
  router.get('/me', authMiddleware.required, authController.me)
}

export default setupAuthRoutes
