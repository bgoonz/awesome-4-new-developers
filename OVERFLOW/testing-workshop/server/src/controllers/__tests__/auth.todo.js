// This is how we'd implement unit tests for the auth controller
// compare this to the integration tests :)
import passport from 'passport'
import {generate} from 'til-server-test-utils'
import db from '../../utils/db'
import * as authController from '../auth.todo'

function setup() {
  const req = {
    body: {},
  }
  const res = {}
  const next = jest.fn()
  Object.assign(res, {
    status: jest.fn(
      function status() {
        return this
      }.bind(res),
    ),
    json: jest.fn(
      function json() {
        return this
      }.bind(res),
    ),
    send: jest.fn(
      function send() {
        return this
      }.bind(res),
    ),
  })
  return {req, res, next}
}

test('username required to register', async () => {
  const {req, res, next} = setup()
  req.body = {password: generate.password()}
  await authController.register(req, res, next)
  expect(res.status).toHaveBeenCalledTimes(1)
  expect(res.status).toHaveBeenCalledWith(422)
  expect(res.json).toHaveBeenCalledTimes(1)
  expect(res.json).toHaveBeenCalledWith({
    errors: {username: expect.any(String)},
  })
  expect(res.json.mock.calls[0][0]).toMatchSnapshot()
})

test('password required to register', async () => {
  const {req, res, next} = setup()
  req.body = {username: generate.username()}
  await authController.register(req, res, next)
  expect(res.status).toHaveBeenCalledTimes(1)
  expect(res.status).toHaveBeenCalledWith(422)
  expect(res.json).toHaveBeenCalledTimes(1)
  expect(res.json).toHaveBeenCalledWith({
    errors: {password: expect.any(String)},
  })
  expect(res.json.mock.calls[0][0]).toMatchSnapshot()
})

test('username required to login', async () => {
  const {req, res, next} = setup()
  req.body = {password: generate.password()}
  await authController.login(req, res, next)
  expect(res.status).toHaveBeenCalledTimes(1)
  expect(res.status).toHaveBeenCalledWith(422)
  expect(res.json).toHaveBeenCalledTimes(1)
  expect(res.json).toHaveBeenCalledWith({
    errors: {username: expect.any(String)},
  })
  expect(res.json.mock.calls[0][0]).toMatchSnapshot()
})

test('password required to login', async () => {
  const {req, res, next} = setup()
  req.body = {username: generate.username()}
  await authController.login(req, res, next)
  expect(res.status).toHaveBeenCalledTimes(1)
  expect(res.status).toHaveBeenCalledWith(422)
  expect(res.json).toHaveBeenCalledTimes(1)
  expect(res.json).toHaveBeenCalledWith({
    errors: {password: expect.any(String)},
  })
  expect(res.json.mock.calls[0][0]).toMatchSnapshot()
})

test('error returned by passport.authenticate will be thrown through', async () => {
  jest.spyOn(passport, 'authenticate')
  const fakeError = new Error('fake error')
  passport.authenticate.mockImplementationOnce((type, options, cb) => {
    cb(fakeError)
  })
  const {req, res, next} = setup()
  req.body = generate.loginForm()
  const error = await authController.login(req, res, next).catch(e => e)
  expect(error).toBe(fakeError)
  passport.authenticate.mockRestore()
})

test('username must be unique', async () => {
  const username = generate.username()
  const {id: userId} = await db.insertUser(generate.userData({username}))

  const {req, res, next} = setup()
  req.body = generate.loginForm({username})
  await authController.register(req, res, next)
  expect(res.status).toHaveBeenCalledTimes(1)
  expect(res.status).toHaveBeenCalledWith(422)
  expect(res.json).toHaveBeenCalledTimes(1)
  expect(res.json).toHaveBeenCalledWith({
    errors: {username: expect.any(String)},
  })
  expect(res.json.mock.calls[0][0]).toMatchSnapshot()

  await db.deleteUser({id: userId})
})

test('me without req.user returns 404', async () => {
  const {req, res, next} = setup()
  await authController.me(req, res, next)
  expect(res.status).toHaveBeenCalledTimes(1)
  expect(res.status).toHaveBeenCalledWith(404)
  expect(res.send).toHaveBeenCalledTimes(1)
  expect(res.send).toHaveBeenCalledWith()
})

test('successful login returns user', async () => {
  jest.spyOn(passport, 'authenticate')
  const loginInfo = generate.loginForm()
  const fakeUser = generate.userData({id: generate.id(), ...loginInfo})
  const fakeToken = generate.token(fakeUser)
  passport.authenticate.mockImplementationOnce((type, options, cb) => {
    cb(null, fakeUser)
  })
  const {req, res, next} = setup()
  req.body = loginInfo
  await authController.login(req, res, next)
  expect(res.status).not.toHaveBeenCalled()
  expect(res.json).toHaveBeenCalledTimes(1)
  expect(res.json).toHaveBeenCalledWith({
    user: {id: fakeUser.id, token: fakeToken, username: fakeUser.username},
  })

  passport.authenticate.mockRestore()
})

test('failed login returns info', async () => {
  jest.spyOn(passport, 'authenticate')
  const fakeInfo = {error: {username: 'is not cool'}}
  passport.authenticate.mockImplementationOnce((type, options, cb) => {
    cb(null, false, fakeInfo)
  })
  const {req, res, next} = setup()
  req.body = generate.loginForm()
  await authController.login(req, res, next)
  expect(res.status).toHaveBeenCalledTimes(1)
  expect(res.status).toHaveBeenCalledWith(422)
  expect(res.json).toHaveBeenCalledTimes(1)
  expect(res.json).toHaveBeenCalledWith(fakeInfo)

  passport.authenticate.mockRestore()
})

test('registering a user inserts the user into the database', async () => {
  const {req, res, next} = setup()
  const loginForm = generate.loginForm()
  req.body = loginForm
  await authController.register(req, res, next)
  const [userInDb] = await db.getUsers(u => u.username === loginForm.username)
  expect(res.status).not.toHaveBeenCalled()
  expect(res.json).toHaveBeenCalledTimes(1)
  expect(res.json).toHaveBeenCalledWith({
    user: {
      id: userInDb.id,
      username: userInDb.username,
      token: generate.token(userInDb),
    },
  })
})

// Here's where you'll add your test to fix the bug!
// The bug lives in the file `../auth.todo` and is fixed
// in the file `../auth` if you need to reference that.
//
// Create an async test to test the authController.me function
// You'll notice that authController.me expects a user object
// on the `req.user` property, so you'll need to generate a
// new user with an id property. Like so:
// generate.userData({id: generate.id()})
// Your test should ensure that res.json was called once and
// only with the right data from the user.
//
// The bug is that the controller is not sending back the token
// which is resulting in authentication not working.
// The response should include the token.
//
// You can verify the token using expect.any(String) or any
// other means you feel bring you confidence this wont break
// again. (There is a generate.token method if you want to try that).

//////// Elaboration & Feedback /////////
// When you've finished with the exercises:
// 1. Copy the URL below into your browser and fill out the form
// 2. remove the `.skip` from the test below
// 3. Change submitted from `false` to `true`
// 4. And you're all done!
/*
http://ws.kcd.im/?ws=Testing&e=postsController&em=
*/
test.skip('I submitted my elaboration and feedback', () => {
  const submitted = false // change this when you've submitted!
  expect(submitted).toBe(true)
})
////////////////////////////////
