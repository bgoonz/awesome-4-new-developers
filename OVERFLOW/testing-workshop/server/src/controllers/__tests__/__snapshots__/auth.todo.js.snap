// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`password required to login 1`] = `
Object {
  "errors": Object {
    "password": "can't be blank",
  },
}
`;

exports[`password required to register 1`] = `
Object {
  "errors": Object {
    "password": "can't be blank",
  },
}
`;

exports[`username must be unique 1`] = `
Object {
  "errors": Object {
    "username": "taken",
  },
}
`;

exports[`username required to login 1`] = `
Object {
  "errors": Object {
    "username": "can't be blank",
  },
}
`;

exports[`username required to register 1`] = `
Object {
  "errors": Object {
    "username": "can't be blank",
  },
}
`;
